crates/ty_python_semantic/src/types.rs

```diff
@@ -8993,16 +8990,136 @@ impl<'db> ConstructorCallError<'db> {
+    /// variable declared as having type `A`.
+    ///
+    /// For a pair of [fully static] types `A` and `B`, the assignability relation
+    /// between `A` and `B` is the same as the subtyping relation.
+    ///
+    /// Between a pair of `C` and `D` where either `C` or `D` is not fully static, the
+    /// assignability relation may be more permissive than the subtyping relation. `D`
+    /// can be said to be assignable to `C` if *any* possibly fully static [materialization]
+    /// of `D` is a subtype of *any* possible fully static materialization of `C`.
+    ///
+    /// For example, `Any` is not a subtype of `int`, because there are possible
+    /// materializations of `Any` (e.g., `str`) that are not subtypes of `int`.
+    /// `Any` is *assignable* to `int`, however, as there are *some* possible materializations
+    /// of `Any` (such as `int` itself!) that *are* subtypes of `int`. `Any` cannot even
+    /// be considered a subtype of itself, as two separate uses of `Any` in the same scope
+    /// might materialize to different types between which there would exist no subtyping
+    /// relation; nor is `Any` a subtype of `int | Any`, for the same reason. Nonetheless,
+    /// `Any` is assignable to both `Any` and `int | Any`.
+    ///
+    /// While `Any` can materialize to anything, the presence of `Any` in a type does not
+    /// necessarily make it assignable to everything. For example, `list[Any]` is not
+    /// assignable to `int`, because there are no possible fully static types we could
+    /// substitute for `Any` in this type that would make it a subtype of `int`. For the
+    /// same reason, a union such as `str | Any` is not assignable to `int`.
+    ///
+    /// [fully static]: https://typing.python.org/en/latest/spec/glossary.html#term-fully-static-type
+    /// [materialization]: https://typing.python.org/en/latest/spec/glossary.html#term-materialize
     Assignability,
+
+    /// The "union simplification" relation.
+    ///
+    /// The union simplification relation dictates whether the union `A | B` can be
+    /// pragmatically simplified to the type `A` without downstream consequences on ty's
+    /// inference of types elsewhere.
+    ///
+    /// For a pair of [fully static] types `A` and `B`, the assignability relation
+    /// between `A` and `B` is the same as the subtyping relation.
+    ///
+    /// Between a pair of `C` and `D` where either `C` or `D` is not fully static, the
+    /// union simplification relation sits in between the subtyping relation and the
+    /// assignability relation. `D` can be said to be redundant in a union with `C` if either:
+    ///
+    /// 1. `D` is a subtype of `C`; or,
+    /// 2. The top materialization of the type `C | D` is equivalent to the top
+    ///    materialization of `C`, *and* the bottom materialization of `C | D` is equivalent
+    ///    to the bottom materialization of `C`.
```

Comment:
I think (1) is redundant :drum: with (2), and you can simplify the whole thing to "`C | D` is gradually equivalent to `C`". That even nicely maps with the intuitive English definition. (You might still keep (1) as a separate check in the implementation, as an optimization, but that's not needed here in the theoretical definition)